flowapi/flowapi/permissions.py

```diff
@@ -1,245 +1,64 @@
-            )
-        )
-
-
-@functools.singledispatch
-def tree_walk_to_scope_list(tree_walk) -> str:
-    """
-
-    Parameters
-    ----------
-    tree_walk : str, list or str or tuple of list of str
-        A path through the tree
-
-    Yields
-    ------
-    str
-        The tree walk as a . delimited string
-
-    """
-    yield tree_walk
+def _flatten_on_key_inner(root, key_of_interest):
+    raise TypeError
 
 
-@tree_walk_to_scope_list.register
-def _(tree_walk: list) -> str:
-    yield ".".join(tree_walk)
-
-
-@tree_walk_to_scope_list.register
-def _(tree_walk: tuple) -> str:
-    for walk in tree_walk:
-        yield from tree_walk_to_scope_list(walk)
-
+@_flatten_on_key_inner.register(dict)
+def _(root, key_of_interest):
+    for node, value in root.items():
+        if is_flat(value):
+            pass
+        else:
+            yield from _flatten_on_key_inner(value, key_of_interest)
+            if node == key_of_interest:
+                # We cannot change the size of a dict mid-iterate, so instead we mark it for
+                # deletion post-iterate
+                root[node] = {}
+                yield value
 
-def per_query_scopes(
-    *, queries: dict, argument_names_to_extract: List[str] = ["aggregation_unit"]
-) -> Iterable[str]:
-    """
-    Constructs and yields query scopes of the form:
-    <query_kind>:<arg_name>:<arg_val>
-    where arg_val may be a query kind, or the name of an aggregation unit if applicable.
 
-    One scope is yielded for each viable query structure, so for queries which contain two child queries
-    two scopes are yielded. If that query has 3 possible aggregation units, then 6 scopes are yielded altogether.
+@_flatten_on_key_inner.register(list)
+def _(root, key_of_interest):
+    for value in root:
+        yield from _flatten_on_key_inner(value, key_of_interest)
 
-    Parameters
-    ----------
-    tree : dict
-        Dict of nested queries
-    all_queries : dict
-        All queries
 
-    Yields
-    ------
-    str
-        Query scope of the form <query_kind>:<arg_name>:<arg_val>
+def _clean_empties(in_dict, marker):
+    out = {}
+    for key, value in in_dict.items():
+        if value != {marker: {}}:
+            out[key] = value
+    return out
 
-    Examples
-    --------
-    >>>list(make_per_query_scopes(queries={"DUMMY": {}}))
-    ["dummy"]
-    >>>list(make_per_query_scopes(queries={"DUMMY": {}}))
-    ["dummy:aggregation_unit:DUMMY_UNIT",]
 
-    """
-    nested_qs = paths_to_nested_dict(
-        queries=queries, argument_names_to_extract=argument_names_to_extract
-    )
-    yield from (
-        "&".join([action, *tree_walk_to_scope_list(scope_list)])
-        for scope_list in valid_tree_walks(nested_qs)
-        for action in ("get_result", "run")
-    )
-    yield "get_result&available_dates"
+def flatten_on_key(in_iter, key, _in_place=False):
+    if not _in_place:
+        in_iter = deepcopy(in_iter)
+    out = list(_flatten_on_key_inner(in_iter, key))
```

Comment:
Why make it into a list here?

Reply:
Logic being that this function is for returning a set of discrete 'things' for processing, a list made sense - I don't think you can make an assumption about what keys are relevant when this is applied. Probably should have been 'split_on_key'.

Reply:
Yeah, but I mean you could just glom the generators together, right? yield from a generator running `_clean_empties` over `_flatten_on_key_inner`?

Reply:
Ah, point - I'll fix that up